includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -1626,7 +1626,7 @@ private function check_attr_spec_rule_value_regex_casei( DOMElement $node, $attr
 	private function check_attr_spec_rule_valid_url( DOMElement $node, $attr_name, $attr_spec_rule ) {
 		if ( isset( $attr_spec_rule[ AMP_Rule_Spec::VALUE_URL ] ) && $node->hasAttribute( $attr_name ) ) {
 			foreach ( $this->extract_attribute_urls( $node->getAttributeNode( $attr_name ) ) as $url ) {
-				$url = urldecode( $url );
+				$url = urldecode( trim( $url ) );
```

Comment:
Note that the AMP validator does this same trimming up front: https://github.com/ampproject/amphtml/blob/af1e3a550feeafd732226202b8d1f26dcefefa18/validator/engine/parse-url.js#L214-L215

However, it is also [doing one more thing](https://github.com/ampproject/amphtml/blob/af1e3a550feeafd732226202b8d1f26dcefefa18/validator/engine/parse-url.js#L217-L218):

```js
    // Browsers remove Tab/CR/LF from the entire URL, so we do too.
    unparsed = unparsed.replace(/[\t\r\n]/g, '');
```

I think that should be incorporated here as well.

Reply:
This being the case, instead of calling `trim()` directly, let's create a new private method like `normalize_url_from_attribute_value()` which does both the trimming and the tab/newline removal and then use that method everywhere that you are currently using `trim()` instead.

Reply:
Sounds good, it'll probably be several hours before I could do that, if that's alright.

Reply:
Sure thing

Reply:
https://github.com/ampproject/amp-wp/pull/4166/commits/911602d04c8fff4f73a382ba9ac0c600e0bdfe1b adds a method to normalize the URLs, sorry for the delay.